r/src/array.cpp

```diff
@@ -0,0 +1,209 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow_types.h"
+
+using namespace Rcpp;
+using namespace arrow;
+
+namespace arrow{
+namespace r{
+
+template <int RTYPE, typename Vec = Rcpp::Vector<RTYPE> >
+class SimpleRBuffer : public arrow::Buffer {
+public:
+
+  SimpleRBuffer(Vec vec) :
+    Buffer(reinterpret_cast<const uint8_t*>(vec.begin()), vec.size() * sizeof(typename Vec::stored_type)),
+    vec_(vec)
+  {}
+
+private:
+  // vec_ holds the memory
+  Vec vec_;
+};
+
+template <int RTYPE, typename Type>
+std::shared_ptr<arrow::Array> SimpleArray(SEXP x){
+
+  Rcpp::Vector<RTYPE> vec(x);
+  std::vector<std::shared_ptr<arrow::Buffer>> buffers {
+    nullptr,
+    std::make_shared<SimpleRBuffer<RTYPE>>(vec)
+  };
+
+  int null_count = 0;
+  if (RTYPE != RAWSXP) {
+    // TODO: maybe first count NA in a first pass so that we
+    //       can allocate only if needed
+    std::shared_ptr<arrow::Buffer> null_bitmap;
+    R_ERROR_NOT_OK(arrow::AllocateBuffer(vec.size(), &null_bitmap));
+
+    auto null_bitmap_data = null_bitmap->mutable_data();
+    for (int i=0; i < vec.size(); i++) {
+      if (Rcpp::Vector<RTYPE>::is_na(vec[i]) ) {
+        BitUtil::SetBit(null_bitmap_data, i);
+        null_count++;
+      } else {
+        BitUtil::ClearBit(null_bitmap_data, i);
+      }
+    }
+    buffers[0] = std::move(null_bitmap);
+  }
+
+  auto data = ArrayData::Make(
+    std::make_shared<Type>(),
+    LENGTH(x),
+    std::move(buffers),
+    null_count,
+    0 /*offset*/
+  );
+
+  // return the right Array class
+  return std::make_shared<arrow::NumericArray<Type>>(data);
+}
+
+}
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::Array> Array__from_vector(SEXP x){
+  switch(TYPEOF(x)){
+    case INTSXP:
+      if (Rf_isFactor(x)) {
+        break;
+      }
+    return arrow::r::SimpleArray<INTSXP, arrow::Int32Type>(x);
+    case REALSXP:
+      // TODO: Dates, ...
+    return arrow::r::SimpleArray<REALSXP, arrow::DoubleType>(x);
+    case RAWSXP:
+      return arrow::r::SimpleArray<RAWSXP, arrow::Int8Type>(x);
+    default:
+      break;
+  }
+
+  stop("not handled");
+  return nullptr;
+}
+
+
+template <int RTYPE>
+inline SEXP simple_Array_to_Vector(const std::shared_ptr<arrow::Array>& array ){
+  using stored_type = typename Rcpp::Vector<RTYPE>::stored_type;
+  auto start = reinterpret_cast<const stored_type*>(array->data()->buffers[1]->data() + array->offset() * sizeof(stored_type));
+
+  size_t n = array->length();
+  Rcpp::Vector<RTYPE> vec(start, start + n);
+  if (array->null_count() && RTYPE != RAWSXP) {
+    // TODO: not sure what to do with RAWSXP since
+    //       R raw vector do not have a concept of missing data
+
+    auto bitmap_data = array->null_bitmap()->data();
+    for (size_t i=0; i < n; i++) {
+      if (BitUtil::GetBit(bitmap_data, i)) {
```

Comment:
Use BitmapReader here for a bit better perf

Reply:
Thanks.

Reply:
Done. also using BitmapWriter for the reverse operation. @wes Is that ok that these are in the `internal` namespace and not marked as `ARROW_EXPORT` ?

Reply:
Yeah, it's OK because they are header-only